### PR TITLE
Download scoreboard

### PR DESCRIPTION
**Pull request by Changron**
can download two type scoreboards

### CODE REVIEW

File: contest/views.py [hunk 1]
```diff
@@ -222,3 +222,11 @@ def reply(request):
                     % (request.user.username, replied_clarification.id))
             return redirect('contest:contest',contest)
     return redirect('contest:archive')
+
+def download(request):
+    what = request.POST.get('type')
+    if what == 'scoreboard':
```
Comment (henryyang42) [inline contest/views.py]: Maybe ``request.method == 'POST' and 'scoreboard' in request.POST`` would be better.
Btw, does download the file require any permission?

Comment (Changron) [inline contest/views.py]: @nickelcheng  do we need to check any permission?

Comment (nickelcheng) [inline contest/views.py]: In my opinion, it doesn't need to check any permission since everyone can see scoreboard.

Comment (Changron) [inline contest/views.py]: Maybe there's something to download in the future. So that I use that coding structure

File: contest/contest_info.py [hunk 2]
```diff
@@ -99,6 +103,72 @@ def get_scoreboard(contest):
 
     return scoreboard
 
+def get_scoreboard_csv(contest_id):
+    try:
```
Comment (nickelcheng) [inline contest/contest_info.py]: Is there a function ```get_contest_or_404```?
If so, I prefer using that function.

File: contest/contest_info.py [hunk 3]
```diff
@@ -99,6 +103,72 @@ def get_scoreboard(contest):
 
     return scoreboard
 
+def get_scoreboard_csv(contest_id):
```
Comment (nickelcheng) [inline contest/contest_info.py]: Is that possible to let user choose which version of scoreboard(penalty or testcase) they want to download?

File: contest/contest_info.py [hunk 4]
```diff
@@ -99,6 +103,72 @@ def get_scoreboard(contest):
...
+    for problem in scoreboard.problems:
+        title.append(problem.id)
+    title.append('Total')
+    writer.writerow(title)
+    #user data
+    for counter, user in enumerate(scoreboard.users):
+        user_row = [counter, user.username]
+        for problem in user.problems:
+            penalty = problem.get_penalty(contest.start_time)
+            user_row.append(penalty)
```
Comment (nickelcheng) [inline contest/contest_info.py]: I prefer using another format:
```submissions before first AC``` / ```first AC time```(in minutes)
if a problem didn't get AC, use ```--``` instead
For example,
if an user got AC for the first submission, and submission time is 10 minutes after contest start, use ```1/10```
if an user had tried 4 times but never got AC, use ```4/--```

File: contest/contest_info.py [hunk 5]
```diff
@@ -99,6 +103,72 @@ def get_scoreboard(contest):
...
+    response = HttpResponse(content_type='text/csv')
+    filename = str(contest.cname) + "-scoreboard"
+    response['Content-Disposition'] = 'attachment; filename=' + filename
+
+    #init
+    writer = csv.writer(response)
+    #penalty scoreboard csv
+    scoreboard.sort_users_by_penalty()
+    #title
+    title = ['Penalty','User']
```
Comment (bruce3557) [inline contest/contest_info.py]: Add a space after ```,```.

File: contest/contest_info.py [hunk 6]
```diff
@@ -99,6 +103,72 @@ def get_scoreboard(contest):
...
+    for counter, user in enumerate(scoreboard.users):
+        user_row = [counter, user.username]
+        for problem in user.problems:
+            penalty = problem.get_penalty(contest.start_time)
+            user_row.append(penalty)
+        total_penalty = user.get_penalty(contest.start_time)
+        user_row.append(total_penalty)
+        writer.writerow(user_row)
+
+    footer = ['Passed','']
```
Comment (bruce3557) [inline contest/contest_info.py]: ditto

File: contest/contest_info.py [hunk 7]
```diff
@@ -99,6 +103,72 @@ def get_scoreboard(contest):
...
+        footer.append(problem.pass_user)
+    writer.writerow(footer)
+
+    #to saparate two scoreboard
+    writer.writerow([])
+
+    #testcases scoreboard csv
+    scoreboard.sort_users_by_solved_testcases()
+    #title
+    title = ['Testcase','User']
```
Comment (bruce3557) [inline contest/contest_info.py]: ditto

File: contest/contest_info.py [hunk 8]
```diff
@@ -99,6 +103,72 @@ def get_scoreboard(contest):
...
+        user_row = [counter,  user.username]
+        for problem in user.problems:
+            passed_testcases = problem.get_testcases_solved()
+            total_testcases = problem.total_testcases
+            user_row.append(str(passed_testcases) + '/' + str(total_testcases))
+        user_total_testcases = user.get_testcases_solved()
+        user_row.append(user_total_testcases)
+        writer.writerow(user_row)
+
+    footer = ['Passed Testcases','']
```
Comment (bruce3557) [inline contest/contest_info.py]: please add a space after ```,```.

File: contest/contest_info.py [hunk 9]
```diff
@@ -99,6 +103,72 @@ def get_scoreboard(contest):
...
+    scoreboard.sort_users_by_solved_testcases()
+    #title
+    title = ['Testcase','User']
+    for problem in scoreboard.problems:
+        title.append(problem.id)
+    title.append('Total')
+    writer.writerow(title)
+    #user data
+    for counter, user in enumerate(scoreboard.users):
+        user_row = [counter,  user.username]
```
Comment (bruce3557) [inline contest/contest_info.py]: remove additional space after ```,```.
Only one needed.

File: contest/contest_info.py (hunk 3 above)
Comment (Changron) [inline contest/contest_info.py]: I'll make two buttons

**Comment by bruce3557** [comment]
+1

File: contest/contest_info.py [hunk 10]
```diff
@@ -107,7 +111,74 @@ def get_scoreboard(contest):
...
+    title.append('Total')
+    writer.writerow(title)
+    #user data
+    for counter, user in enumerate(scoreboard.users):
+        user_row = [counter+1, user.username]
+        for problem in user.problems:
+            penalty = problem.get_penalty(contest.start_time)
+            user_row.append(penalty)
+        total_penalty = user.get_penalty(contest.start_time)
+        user_row.append(total_penalty)
```
Comment (nickelcheng) [inline contest/contest_info.py]: I prefer using another format:
```submissions before first AC``` / ```first AC time```(in minutes)
if a problem didn't get AC, use ```--``` instead
For example,
if an user got AC for the first submission, and submission time is 10 minutes after contest start, use ```1/10```
if an user had tried 4 times but never got AC, use ```4/--```

File: contest/contest_info.py [hunk 11]
```diff
@@ -107,7 +111,74 @@ def get_scoreboard(contest):
...
+    elif scoreboard_type == "testcases":
+        write_scoreboard_csv_testcases(writer, contest, scoreboard)
+
+    return response
+
+def write_scoreboard_csv_penalty(writer, contest, scoreboard):
+    #penalty scoreboard csv
+    scoreboard.sort_users_by_penalty()
+    #title
+    title = ['Penalty', 'User']
```
Comment (nickelcheng) [inline contest/contest_info.py]: I prefer to use ```Rank``` instead of ```Penalty```.

File: contest/contest_info.py [hunk 12]
```diff
@@ -107,7 +111,74 @@ def get_scoreboard(contest):
...
+    footer = ['Passed', '']  
+    for problem in scoreboard.problems:
+        footer.append(problem.pass_user)
+    writer.writerow(footer)
+
+def write_scoreboard_csv_testcases(writer, contest, scoreboard):
+    #testcases scoreboard csv
+    scoreboard.sort_users_by_solved_testcases()
+    #title
+    title = ['Testcase', 'User']
```
Comment (nickelcheng) [inline contest/contest_info.py]: I prefer to use ```Rank``` instead of ```Testcase```.

**Comment by nickelcheng** [comment]
+1 besides comments.

**Comment by Changron** [comment]
fixed

**Pull request merged**